### ISSUE TITLE
Expected type is not propagated through branching

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/haxe#9450]
Could the following code be made working? It would be nice :)

```haxe
function get() {
	return new Map<Int,Int>();
}

function getE() {
	return E.A;
}

enum E { A; B; }
enum E2 { A; B; }

function main() {
	// doesn't compile
	trace(try get() catch (e) []); // haxe.ds.Map<Int, Int> should be Array<Unknown<0>>
	trace(if (Math.random() > 0.5) get() else []); // Array<Unknown<0>> should be haxe.ds.Map<Int, Int>
	trace(if (Math.random() > 0.5) getE() else A); // E2 should be E
}
```

Also, in the `try` example it complains about `get()` rather than `[]`, which is confusing because it doesn't follow syntatical order.

**Comment by nadako** [comment]
I get it though that it should NOT complain about expected type mismatch if these expressions are not used as a value, and we still want type "minification", which seems to be a bit incompatible with expected type, so changing this might be quite involved.

**Comment by Simn** [comment]
The whole idea behind finding the minimal type is to _not_ type the second branch like the first one...

**Comment by nadako** [comment]
Yeah, I get that... But maybe it's theoretically possible to first type a branch against the first type, if that fails - try parent type, if that succeeds - set branch type to that parent type and proceed to the next branch. Something like that, just thinking...

**Comment by nadako** [comment]
> Also, in the `try` example it complains about `get()` rather than `[]`, which is confusing because it doesn't follow syntatical order.

Anyway, this one looks like it could/should be fixed anyway? I guess there's some List.rev missing?

**Comment by nadako** [comment]
> Yeah, I get that... But maybe it's theoretically possible to first type a branch against the first type, if that fails - try parent type, if that succeeds - set branch type to that parent type and proceed to the next branch. Something like that, just thinking...

I guess this is also tricky because of monomorph binding and no way to cancel that (same issue as with overloading)?

**Comment by back2dos** [comment]
> The whole idea behind finding the minimal type is to not type the second branch like the first one...

But if the first branch is an abstract/enum, does this really make sense to expect anything other in the second? I mean yeah, one could determine `if (Math.random() > 0.5) getE() else A` to be `EnumValue`, but that doesn't seem horribly useful

**Comment by Simn** [comment]
I don't think we should make typing exceptions for this case.

**Comment by Simn** [comment]
Uhm, I just tried your sample and it isn't failing for me...

**Comment by nadako** [comment]
Wait what... That fails for me for `-js`, but doesn't fail for e.g. `--interp` or `-swf` O_o

**Comment by nadako** [comment]
But also (with `--interp`):

![image](https://user-images.githubusercontent.com/49749/83728570-d7671500-a646-11ea-88c1-76e61d8cab0e.png)

**Comment by nadako** [comment]
Or even without the compilation server:

![image](https://user-images.githubusercontent.com/49749/83728627-f1085c80-a646-11ea-9f7b-83d819f2529d.png)

**Comment by nadako** [comment]
Same behaviour for `-swf`...

**Comment by RealyUniqueName** [comment]
This is reproducible for me too.

**Comment by nadako** [comment]
Well ok the different printing is of course expected because of Math.random, I shouldn't post before coffee. But still why the type check is failing on one platform but doesn't fail on another.
